Sources/PackageGet.php

```diff
@@ -666,19 +666,16 @@ function PackageUpload()
 
 	// Make sure it has a sane filename.
 	$_FILES['package']['name'] = preg_replace(array('/\s/', '/\.[\.]+/', '/[^\w_\.\-]/'), array('_', '.', ''), $_FILES['package']['name']);
-	$extension = substr(strrchr(strtolower($_FILES['package']['name']), '.'), 0);
-	if (!in_array($extension, $allowext))
-	{
+
+	$found_ext = preg_match('/\.(zip|tgz|tar\.gz)$/i', $_FILES['package']['name'], $match);
+	if ($found_ext === 0)
 		fatal_lang_error('package_upload_error_supports', false, array('zip, tgz, tar.gz'));
-	}
 
 	// We only need the filename...
-	$extension = ($extension == '.gz') ? '.tar.gz' : $extension;
-	$packageName = time() . $extension;
+	$packageName = basename($_FILES['package']['name']);
```

Comment:
```suggestion
	$packageName = substr($_FILES['package']['name'], 0, -strlen($match[0]));
```

Reply:
Don't trust `basename()`?

Reply:
I trust it to give us, e.g., `derp.tar.gz`. But in `package_unique_filename()`, the `$filename` variable was being handled as if it contained only `derp`.

Reply:
To be clear, the version of `package_unique_filename()` in c531cfd was first checking if `packages/derp.tar.gztar.gz` existed, and then if `packages/derp.tar.gz_1tar.gz` existed, before attempting to rename to `derp.tar.gz_2tar.gz`

Reply:
Perhaps you were thinking of `pathinfo($_FILES['package']['name'], PATHINFO_FILENAME)`? That would in theory have provided what `package_unique_filename()` seems to expect, except that it would only remove `.gz` instead of `.tar.gz` from the end of the name.

Reply:
I didn't use `pathinfo()` for that very reason. Hence the regex.